packages/firestore/test/unit/specs/bundle_spec.test.ts

```diff
@@ -0,0 +1,357 @@
+      const bundleString1 = bundleWithDocument({
+        key: docA.key,
+        readTime: 500,
+        createTime: 250,
+        updateTime: 500,
+        content: { key: { stringValue: 'b' } }
+      });
+
+      const bundleString2 = bundleWithDocument({
+        key: docA.key,
+        readTime: 1001,
+        createTime: 250,
+        updateTime: 1001,
+        content: { key: { stringValue: 'fromBundle' } }
+      });
+      return (
+        spec()
+          .withGCEnabled(false)
+          .userListens(query)
+          .watchAcksFull(query, 250, docA)
+          .expectEvents(query, {
+            added: [doc('collection/a', 250, { key: 'a' })]
+          })
+          .userPatches('collection/a', { key: 'patched' })
+          .expectEvents(query, {
+            modified: [
+              doc(
+                'collection/a',
+                250,
+                { key: 'patched' },
+                { hasLocalMutations: true }
+              )
+            ],
+            hasPendingWrites: true
+          })
+          .writeAcks('collection/a', 1000)
+          // loading bundleString1 will not raise snapshots, because it is before
+          // the acknowledged mutation.
+          .loadBundle(bundleString1)
+          // loading bundleString2 will raise a snapshot, because it is after
+          // the acknowledged mutation.
+          .loadBundle(bundleString2)
+          .expectEvents(query, {
+            modified: [doc('collection/a', 1001, { key: 'fromBundle' })]
+          })
+      );
+    }
+  );
+
+  specTest(
+    'Newer docs from bundles should keep not raise snapshot if there are unacknowledged writes.',
+    [],
+    () => {
+      const query = Query.atPath(path('collection'));
+      const docA = doc('collection/a', 250, { key: 'a' });
+
+      const bundleString1 = bundleWithDocument({
+        key: docA.key,
+        readTime: 500,
+        createTime: 250,
+        updateTime: 500,
+        content: { key: { stringValue: 'b' } }
+      });
+
+      const bundleString2 = bundleWithDocument({
+        key: docA.key,
+        readTime: 1001,
+        createTime: 250,
+        updateTime: 1001,
+        content: { key: { stringValue: 'fromBundle' } }
+      });
+
+      return (
+        spec()
+          .withGCEnabled(false)
+          .userListens(query)
+          .watchAcksFull(query, 250, docA)
+          .expectEvents(query, {
+            added: [doc('collection/a', 250, { key: 'a' })]
+          })
+          .userPatches('collection/a', { key: 'patched' })
+          .expectEvents(query, {
+            modified: [
+              doc(
+                'collection/a',
+                250,
+                { key: 'patched' },
+                { hasLocalMutations: true }
+              )
+            ],
+            hasPendingWrites: true
+          })
+          // Loading both bundles will not raise snapshots, because of the
+          // mutation is not acknowledged.
+          .loadBundle(bundleString1)
```

Comment:
This bundle is ignored (as tested above), so we probably don't need to test it here. We could also add a small separate test that just verifies this code path.